### DISCUSSION TITLE
Can BackgroundTasks be used inside APIRoute class in FastAPI for logging purpose?

### DISCUSSION THREAD

**Discussion by indigo-tribe** [fastapi/fastapi#9532] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
def log_response(request, response, trace_id, start_time):
    duration = (time.time() - start_time) * 1000
    formatted_duration = f"{duration:.2f}"
    logger.info(
        f"Request-Response: {request.method} {request.url.path} - {response.status_code} | {formatted_duration} ms"
    )

class LogRoute(APIRoute):
    def get_route_handler(self) -> Callable:
        original_route_handler = super().get_route_handler()

        async def custom_route_handler(request: Request, background_tasks: BackgroundTasks) -> Response:
            trace_id = str(uuid.uuid4().hex)
            with logger.contextualize(trace_id=trace_id):
                try:
                    start_time = time.time()
                    response: Response = await original_route_handler(request)
                    background_tasks.add_task(
                        log_response, request, response, trace_id, start_time
                    )
                    return response
                except Exception:
                    logger.error(f"Error processing request: {sys.exc_info()}")
                    raise

        return custom_route_handler
```

### Description

Hi,

In the mentioned code, I am trying to perform request/ response logging through BackgroundTasks. Instead of repeating the same code for BackgroundTasks inside each end-point, I decided to create an APIRoute class and use this inside each router. But, this is giving the below error:

TypeError: LogRoute.get_route_handler.<locals>.custom_route_handler() missing 1 required positional argument: 'background_tasks'

Please help to resolve this issue. Also, please suggest if there is a better way to achieve what I am trying to do.

### Operating System

macOS

### FastAPI Version

0.95.1

### Python Version

3.11.3

**Comment by indigo-tribe** [comment]
Can anyone help me with an answer or an alternative solution?

**Reply by furqanalishah** [reply]
Hi @indigo-tribe 
How about adding middleware to handle this kind of situation?

**Reply by indigo-tribe** [reply]
@furqanalishah - Initially I tried with middleware only. But getting the json response inside the middleware seems to be a tedious task and I found somewhere on the internet that it may have some performance impacts as well, for what I'm trying to do.

**Comment by furqanalishah** [comment]
The `BackgroundTasks` object is not properly passed and is that's why throwing the error. I mean the BackgroundTasks dependency needs to be properly injected. via ` = BackgroundTasks()`
This below code snippet will hopefully solve the issue. 
```
async def custom_route_handler(request: Request, background_tasks: BackgroundTasks = BackgroundTasks()) -> Response:
```

**Reply by indigo-tribe** [reply]
Thank you for the response @furqanalishah 
Injecting BackgroundTasks dependency as suggested by you removed the error, but seems like it is not doing what is expected. i.e. the log_response function is not being called. I tried printing inside the log_response function and nothing gets printed.

Here's the modified code:

```python
def log_response(request, response, trace_id, start_time):
    print("INFO: Entering log_response function...")
    duration = (time.time() - start_time) * 1000
    formatted_duration = f"{duration:.2f}"
    logger.info(
        f"Request-Response: {request.method} {request.url.path} - {response.status_code} | {formatted_duration} ms"
    )


class LogRoute(APIRoute):
    def get_route_handler(self) -> Callable:
        original_route_handler = super().get_route_handler()

        async def custom_route_handler(request: Request, background_tasks: BackgroundTasks = BackgroundTasks()) -> Response:
            trace_id = str(uuid.uuid4().hex)
            with logger.contextualize(trace_id=trace_id):
                try:
                    start_time = datetime.now()
                    response: Response = await original_route_handler(request)
                    background_tasks.add_task(log_response, request, response, trace_id, start_time)
                    return response
                except Exception:
                    logger.error(f"Error processing request: {sys.exc_info()}")
                    raise

        return custom_route_handler
```

**Reply by bogh** [reply]
@indigo-tribe 
Could it be that your problem is the `logger.contextualize`?

That context will exit, and after the background task is set to execute. Can you test without the context?

**Comment by matthewbal** [comment]
I just struggled with background tasks today. After trying both FastAPI and Starlette's helpers, I settled on just using pure asyncio - and it seems to be working really well and is much simpler to read/reason about. Here's the code I used - in my case I was demoing using an async database connection but you could just as easily call `asyncio.create_task` from your function.

A gotcha was awaiting the task in the tests - the test itself is a task so you need to await all other tasks except the current one otherwise it will just wait for itself forever.

```
async def async_dummy_object(new_obj: Dummy, db: AsyncSession):
    """
    Create a dummy object after 1 second
    """

    await asyncio.sleep(1)

    logger.debug(f"Inserting new dummy data: {new_obj.dummy_data}")

    db.add(new_obj)
    await db.commit()


@router.post("/in-background", response_model=str)
async def post_dummy_delay(
    payload: DummyInput,
    db: AsyncSession = Depends(get_async_db),
) -> str:
    """
    Insert some data into dummy table, demos using a background task
    """

    new_dummy = Dummy(dummy_data=payload.dummy_data)

    asyncio.create_task(async_dummy_object(new_obj=new_dummy, db=db))

    return "Creating dummy object in background task"
    
 
async def test_dummy_in_background_adds_data(async_client: CustomTestClient, test_async_db: AsyncSession):
    # Given we have a valid payload
    payload = {"dummy_data": "Some Data"}

    # When we post it to the endpoint
    res_data = await async_client.api("/dummy/in-background", json=payload)

    # Then we should get back the message
    assert res_data == "Creating dummy object in background task"

    # And then we should not have a row in the db yet
    count = await test_async_db.scalar(select(func.count()).select_from(mdl.Dummy))
    assert count == 0

    # When we wait until the background task is done
    await asyncio.gather(*asyncio.all_tasks() - {asyncio.current_task()})

    # Then we should have a row in the db
    count = await test_async_db.scalar(select(func.count()).select_from(mdl.Dummy))
    assert count == 1

```

**Reply by nymous** [reply]
Be careful with `asyncio.create_task()`, if you don't save a reference to the task for as long as it needs to live it can be garbage-collected before it's complete. See [this Textual blog](https://textual.textualize.io/blog/2023/02/11/the-heisenbug-lurking-in-your-async-code/) fore more info.

**Reply by matthewbal** [reply]
@nymous damn that's a really good point. Hmmm, I don't know where I could save a reference to the task, since I need the endpoint to return and exit.

I tried background tasks from FastAPI and Starlette and no matter what I did, my endpoint would wait until the task completed before returning. I'm using newest versions of both.

**Reply by matthewbal** [reply]
Hmm, I threw in `gc.collect()` in the endpoint & also before awaiting the tasks to force garbage collection and the task is still always being run, this is very hard to reproduce in tests. 

I have noticed another issue with my approach though - I can't create 2 tasks from the same endpoint because the database connection gets messed up.

**Comment by illpack** [comment]
Hi there, I am also struggling with this issue, I did not manage to make it work using a custom `APIRoute`, but I think it could be done using a Middleware as proposed in this thread... I tried a few whings but I don't manage to get the BackgroundTasks object to execute, neither as a middleware nor as a custom APIRoute. Could any of you solve this? Thank you,

**Comment by sxooler** [comment, marked as answer]
Background tasks are part of the response if I am not mistaken. Shouldn't you do sth like this?

```python
class LogRoute(APIRoute):
    def get_route_handler(self) -> Callable:
        original_route_handler = super().get_route_handler()

        async def custom_route_handler(request: Request) -> Response:
            trace_id = str(uuid.uuid4().hex)
            with logger.contextualize(trace_id=trace_id):
                try:
                    start_time = time.time()
                    response: Response = await original_route_handler(request)
                    background_tasks = BackgroundTasks()
                    background_tasks.add_task(
                        log_response, request, response, trace_id, start_time
                    )
                    response.background = background_tasks
                    return response
                except Exception:
                    logger.error(f"Error processing request: {sys.exc_info()}")
                    raise

        return custom_route_handler
```

It's also a good practice to check `response.background` before assignment as there might be already some tasks.
